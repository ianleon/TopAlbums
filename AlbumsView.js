/**
 * AlbumsView Screen
 *
 * @format
 * @flow strict-local
 */
import React, { Component } from 'react'
import {
  SafeAreaView,
  StyleSheet,
  ScrollView,
  View,
  Button,
  Text,
  StatusBar,
  FlatList,
  Image,
  Linking,
  Pressable
} from 'react-native';

class AlbumsView extends Component {

	state = {
      topAlbums: [],
    };

  	componentDidMount = () => {

		fetch("https://itunes.apple.com/us/rss/topalbums/limit=100/json")
		.then(response => response.json())
		.then(musicData => {

		  let cleanData = musicData.feed.entry.map((ent,i) => {

			  let key = i + "";
			  let position = i+1;
			  let title = ent["im:name"].label;

			  let artist = ent["im:artist"].label;
			  let releaseDate = ent["im:releaseDate"].attributes.label;
			  let category = ent.category.attributes.label;
			  let image = ent["im:image"][0].label;
			  let link = ent.link.attributes.href;

			  return ({key, position, image, releaseDate, title, artist,category, link});
		  });
		  this.setState({topAlbums: cleanData});

		});
    }

	render = () => {
		const { navigate } = this.props.navigation;

		return (<FlatList data={this.state.topAlbums} renderItem={ ({item}) =>
				<SafeAreaView>
					<Pressable onPress={() => navigate('AlbumDetails', { item })}
					style={{ alignItems: "center", flexDirection: 'row', backgroundColor: "#fff", marginBottom: 1 }}>
						<Image style={{ width: 70, height: 70 }} source={{ uri: item.image }} />
						<Text style={styles.listTitles}>{item.title}</Text>
					</Pressable>
				</SafeAreaView>} />);
	}
}

const styles = StyleSheet.create({
  listTitles: {
	padding: 10,
	fontSize: 20,
	flex: 1

  },
});

export default AlbumsView;
