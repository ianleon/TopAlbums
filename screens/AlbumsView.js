/**
 * AlbumsView Screen
 *
 * @format
 * @flow strict-local
 */
import React, { Component } from 'react'
import {
  SafeAreaView,
  StyleSheet,
  ScrollView,
  View,
  Button,
  Text,
  UIManager,
  LayoutAnimation,
  StatusBar,
  FlatList,
  Image,
  Linking,
  Pressable
} from 'react-native';


UIManager.setLayoutAnimationEnabledExperimental &&
UIManager.setLayoutAnimationEnabledExperimental(true);

const ListRow = ({item}) => {
	return <SafeAreaView>
		<Pressable
		onPress={() => navigate('AlbumDetails', {item})}
		style={styles.listRows}>
			<Image style={styles.listImage} source={{ uri: item.image }} />
			<Text style={styles.listTitles}>{item.title}</Text>
		</Pressable>
	</SafeAreaView>
}

class AlbumsView extends Component {

	state = {
      topAlbums: [],
    };


	setAnimation = () => {
		LayoutAnimation.configureNext({
			duration: 500,
			create: {
				type: LayoutAnimation.Types.easeOut,
				property: LayoutAnimation.Properties.scaleY,
			},
		});
	};

  	componentDidMount = () => {

		fetch("https://itunes.apple.com/us/rss/topalbums/limit=100/json")
		.then(response => response.json())
		.then(musicData => {

		  let cleanData = musicData.feed.entry.map((ent,i) => {

			  let key = i + "";
			  let position = i+1;
			  let title = ent["im:name"].label;

			  let artist = ent["im:artist"].label;
			  let releaseDate = ent["im:releaseDate"].attributes.label;
			  let category = ent.category.attributes.label;
			  let image = ent["im:image"][0].label;
			  let link = ent.link.attributes.href;

			  return ({
				  key, position, image, releaseDate,
				  title, artist,category, link});
		  });
		  this.setAnimation();
		  this.setState({topAlbums: cleanData});

		});
    }

	render = () => {
		const { navigate } = this.props.navigation;
		return (<FlatList data={this.state.topAlbums} renderItem={ListRow} />);
	}
}

const styles = StyleSheet.create({
  listTitles: {
	padding: 10,
	fontSize: 20,
	flex: 1
  },
  listImage: {
	  width: 70,
	  height: 70
  },
  listRows: {
	  alignItems: "center",
	  flexDirection: 'row',
	  backgroundColor: "#fff",
	  marginBottom: 1
  }
});

export default AlbumsView;
