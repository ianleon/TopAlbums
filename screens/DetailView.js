/**
 * DetailsView Screen
 *
 * @format
 * @flow strict-local
 */
import React, { Component, useState } from 'react'
import {
  SafeAreaView, StyleSheet, ScrollView,
  View, Button, Text, FlatList, Image,
  Linking, Pressable
} from 'react-native';

import Icon from 'react-native-vector-icons/AntDesign';

const Hairline = () => {
	return <View style={styles.hairline}></View>
}

function StarRatings({stars,onPress}) {

	var starSize = 40;
	var icons = [];
	for (let i = 0; i < 5; i++) {

		icons.push(
			<Icon
			key={i}
			size={starSize}
			name={i < stars ? "star" : "staro"}
			onPress={ () => { onPress(i + 1) } } />
		);
	}

	return <View style={{ flexDirection: "row" }}>{icons}</View>;
}

function AppleMusicLogo({size}) {

  return (<>
    <Icon name="apple1" size={size} color="#FFF" style={styles.amIcon} />
    <Text style={{...styles.amType, fontSize: size}}>MUSIC</Text>
  </>);
}

function AppleMusicButton({link}) {
  return (
    <Pressable onPress={ ()=> Linking.openURL(link) } style={styles.linkStyle}>
      <View style={{ flexDirection: 'row' }}>
        <AppleMusicLogo size={20} />
			</View>
		</Pressable>
  );
}

function DetailView({ route }) {

  const { item : {
    key, position, image, title,
    artist, category, link, releaseDate
  } } = route.params;

  // Using hooks for ratings
  const [rating, setRating] = useState(3);

  return <>
  <ScrollView style={{ flex: 1, flexDirection: 'column', padding: 16}}>
  	<SafeAreaView>

  		<View style={styles.detailHeader}>
  			<Text style={styles.detailTitle}>{ title }</Text>
  			<Image style={{ width: 70, height: 70 }} source={{ uri: image }} />
  		</View>

  		<Text style={styles.detailLabel}>POSITION</Text>
  		<Text style={styles.detailValue}>{ position }</Text>
  		<Hairline />
  		<Text style={styles.detailLabel}>ARTIST</Text>
  		<Text style={styles.detailValue}>{ artist }</Text>
  		<Hairline />
  		<Text style={styles.detailLabel}>GENRE</Text>
  		<Text style={styles.detailValue}>{ category }</Text>
  		<Hairline />
  		<Text style={styles.detailLabel}>RELEASE DATE</Text>
  		<Text style={styles.detailValue}>{ releaseDate }</Text>

  		<StarRatings stars={rating} onPress={(i) => { setRating(i); }}/>

      <AppleMusicButton link={link} />

  	</SafeAreaView>
  </ScrollView>
  </>;
};
const styles = StyleSheet.create({

  amIcon: {
    top: 1,
    marginRight: 5
  },
  amType: {
    color: "white",
    fontWeight: "bold"
  },
  hairline: {
    backgroundColor: "#ddd",
    width: "100%",
    height: 1,
    marginBottom: 25,
  },
  detailHeader: {
    flex: 1,
    flexDirection: 'row',
    justifyContent: "space-between",
    width: "100%",
    paddingBottom: 30
  },
  detailTitle: {
    fontSize: 27,
    paddingRight: 20,
    fontWeight: "bold",
    maxWidth: "80%",
  },
  detailLabel: {
  	fontSize: 15,
  },
  detailValue: {
  	fontSize: 22,
	  paddingBottom: 20
  },
  linkStyle: {
    backgroundColor: "#000",
    borderRadius: 20,
    alignItems: "center",
    padding: 20,
    marginTop: 40,
  },
});

export default DetailView;
