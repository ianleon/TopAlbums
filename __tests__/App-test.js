/**
 * @format
 */

import 'react-native';
import React from 'react';
import {
  SafeAreaView,
  StyleSheet,
  ScrollView,
  View,
  Button,
  Text,
  FlatList,
  Image,
  Linking,
  Pressable
} from 'react-native';
import DetailView from '../DetailView';
import AlbumsView from '../AlbumsView';

// Note: test renderer must be required after react-native.
import renderer from 'react-test-renderer';

it('renders details correctly', () => {

	let testRoute = {
		"params": {
			"item": {
				key: "0",
				position: "1",
				image: "http://example.com",
				title: "Test",
				artist: "Test",
				category: "TestMusic",
				link: "http://example.com",
				releaseDate: "Test"
			}
		}
	};
  	let detailRender = renderer.create(<DetailView route={testRoute} />).toJSON();
  	expect(detailRender).toMatchSnapshot();
});

it('renders top albums correctly', () => {

	let testNavigation = {
		navigation: {
			navigate: () => {}
		}
	};
  	let albumsRender = renderer.create(<AlbumsView navigation={testNavigation} fetch="{}"/>).toJSON();
	expect(albumsRender).toMatchSnapshot();
});
