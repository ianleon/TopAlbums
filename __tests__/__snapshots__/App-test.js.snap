// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders details correctly 1`] = `
<RCTScrollView
  style={
    Object {
      "flex": 1,
      "flexDirection": "column",
      "padding": 16,
    }
  }
>
  <View>
    <RCTSafeAreaView
      emulateUnlessSupported={true}
    >
      <View
        style={
          Object {
            "flex": 1,
            "flexDirection": "row",
            "justifyContent": "space-between",
            "paddingBottom": 30,
            "width": "100%",
          }
        }
      >
        <Text
          style={
            Object {
              "fontSize": 27,
              "fontWeight": "bold",
              "maxWidth": "80%",
              "paddingRight": 20,
            }
          }
        >
          Test
        </Text>
        <Image
          source={
            Object {
              "uri": "http://example.com",
            }
          }
          style={
            Object {
              "height": 70,
              "width": 70,
            }
          }
        />
      </View>
      <Text
        style={
          Object {
            "fontSize": 15,
          }
        }
      >
        POSITION
      </Text>
      <Text
        style={
          Object {
            "fontSize": 22,
            "paddingBottom": 20,
          }
        }
      >
        1
      </Text>
      <View
        style={
          Object {
            "backgroundColor": "#ddd",
            "height": 1,
            "marginBottom": 25,
            "width": "100%",
          }
        }
      />
      <Text
        style={
          Object {
            "fontSize": 15,
          }
        }
      >
        ARTIST
      </Text>
      <Text
        style={
          Object {
            "fontSize": 22,
            "paddingBottom": 20,
          }
        }
      >
        Test
      </Text>
      <View
        style={
          Object {
            "backgroundColor": "#ddd",
            "height": 1,
            "marginBottom": 25,
            "width": "100%",
          }
        }
      />
      <Text
        style={
          Object {
            "fontSize": 15,
          }
        }
      >
        GENRE
      </Text>
      <Text
        style={
          Object {
            "fontSize": 22,
            "paddingBottom": 20,
          }
        }
      >
        TestMusic
      </Text>
      <View
        style={
          Object {
            "backgroundColor": "#ddd",
            "height": 1,
            "marginBottom": 25,
            "width": "100%",
          }
        }
      />
      <Text
        style={
          Object {
            "fontSize": 15,
          }
        }
      >
        RELEASE DATE
      </Text>
      <Text
        style={
          Object {
            "fontSize": 22,
            "paddingBottom": 20,
          }
        }
      >
        Test
      </Text>
      <View
        accessible={true}
        focusable={true}
        onBlur={[Function]}
        onClick={[Function]}
        onFocus={[Function]}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={
          Object {
            "alignItems": "center",
            "backgroundColor": "#000",
            "borderRadius": 20,
            "marginTop": 40,
            "padding": 20,
          }
        }
      >
        <Text
          style={
            Object {
              "color": "white",
              "fontSize": 20,
              "fontWeight": "bold",
            }
          }
        >
          Get
        </Text>
      </View>
    </RCTSafeAreaView>
  </View>
</RCTScrollView>
`;

exports[`renders top albums correctly 1`] = `
<RCTScrollView
  data={Array []}
  disableVirtualization={false}
  getItem={[Function]}
  getItemCount={[Function]}
  horizontal={false}
  initialNumToRender={10}
  keyExtractor={[Function]}
  maxToRenderPerBatch={10}
  onContentSizeChange={[Function]}
  onEndReachedThreshold={2}
  onLayout={[Function]}
  onMomentumScrollEnd={[Function]}
  onScroll={[Function]}
  onScrollBeginDrag={[Function]}
  onScrollEndDrag={[Function]}
  removeClippedSubviews={false}
  renderItem={[Function]}
  scrollEventThrottle={50}
  stickyHeaderIndices={Array []}
  updateCellsBatchingPeriod={50}
  viewabilityConfigCallbackPairs={Array []}
  windowSize={21}
>
  <View />
</RCTScrollView>
`;
